design/004-external-data-handling/004-external-data-handling.md

```diff
@@ -0,0 +1,255 @@
+a sidecar, we could likely achive dynamic mounting of data sources, which would
+be fantastic for data exploration.
+
+### Commands
+
+#### Adding external data:
+
+The examples below are based on the existing `dataset` commands, but we might
+need to envision something new like `renku data source` which the `dataset`
+commands could reference.
+
+
+```
+renku dataset add s3://server/path/to/bucket
+Access Key: 1234
+Secret Key: abcd
+Store credentials? (y/N): y
+Credentials stored in ~/.renku/renku.ini
+```
+
+`s3://` is used here but we should also support at least `nfs://`. Renku should
+ask for credentials in a way that is appropriate for the type of resource that
+is being added.
+
+#### Getting information about the dataset:
+
+```
+renku dataset show very-large-data
+
+Name: very-large-data
+Created: 2022-03-30 15:55:59+02:00
+Creator(s): Rok Roškar <rok.roskar@sdsc.ethz.ch>
+Title: very-large-data
+Description: Data that is very large. Indeed.
+External Sources:
+    - type: s3
+      source: s3://server/path/to/bucket
+      location: None
+```
+
+#### Staging external data:
+
+External data has to somehow be brought into the system where Renku is running.
+This seems to be a "staging" operation, but we could also call it something
+else. "Mounting" is another option but seems too linux-specific.
+
+```
+renku dataset stage very-large-data
+```
+
+This command uses Fuse when necessary to mount data locally. Once it's staged,
+it can be found locally and used in workflows:
+
+```
+renku dataset show very-large-data
+
+Name: very-large-data
+Created: 2022-03-30 15:55:59+02:00
+Creator(s): Rok Roškar <rok.roskar@sdsc.ethz.ch>
+Title: very-large-data
+Description: Data that is very large. Indeed.
+External Sources:
+    - type: s3
+      source: s3://server/path/to/<bucket>
+      location: .renku/mounts/<bucket>
+```
+
+The data is accessible under `data/very-large-data/<bucket>` in the project's directory.
+
+If the data is already mounted by some other means, `dataset stage` can be used
+to let Renku know about it:
+
+```
+renku dataset stage --existing s3://server/path/to/<bucket>:/mounts/bucket
+```
+
+#### Streaming external data
+
+The renku-python API should offer some support for more easily accessing remote
+data, especially for sources from which it is common for data to be processed as
+a stream. Ideally, Renku could lazily stage the external data if access to the
+resource is requested.
+
+## Drawbacks
+
+* Relying on fuse is tricky because it requires elevated privileges on all
+  systems. On linux, the user must have sudo and it isn't clear how this will
+  work at all on windows. Fuse has the potentially to work very seamlessly but
+  can also cause problems. *EDIT*: this turns out not to be an issue; outside of
+  docker containers FUSE does not require additional privileges.
+
+* It might also be inefficient to keep track of which data is actually external
+  when recording/handling workflows.
+
```

Comment:
```suggestion

## Out of Scope
* Saving credentials, so that RenkuLab users don't have to re-enter their credentials at the start of each session.

```